src/instrumentation/do-storage.ts

```diff
@@ -41,25 +63,116 @@ const StorageAttributes: Record<string | symbol, ExtraAttributeFn> = {
 			}
 		}
 		if (argArray.length > 1) {
-			Object.assign(attrs, argArray[1])
+			const options = argArray[1] as DurableObjectPutOptions
+			if ('allowConcurrency' in options) {
+				attrs['db.cf.do.allow_concurrency'] = options.allowConcurrency
+			}
+			if ('noCache' in options) {
+				attrs['db.cf.do.no_cache'] = options.noCache
+			}
 		}
 		return attrs
 	},
 	list(argArray, result: Map<string, unknown>) {
-		// list may be called with no arguments
 		const attrs: Attributes = {
 			'db.cf.do.number_of_results': result.size,
+		} as Record<string, string | number | boolean | undefined>
+		if (argArray.length > 0) {
+			const options = argArray[0] as DurableObjectPutOptions
+			if ('allowConcurrency' in options) {
+				attrs['db.cf.do.allow_concurrency'] = options.allowConcurrency
+			}
+			if ('noCache' in options) {
+				attrs['db.cf.do.no_cache'] = options.noCache
+			}
+			if ('start' in options) {
+				attrs['db.cf.do.start'] = options.noCache
+			}
+			if ('startAfter' in options) {
+				attrs['db.cf.do.start_after'] = options.noCache
+			}
+			if ('end' in options) {
+				attrs['db.cf.do.end'] = options.noCache
+			}
+			if ('prefix' in options) {
+				attrs['db.cf.do.prefix'] = options.noCache
+			}
+			if ('reverse' in options) {
+				attrs['db.cf.do.reverse'] = options.noCache
+			}
+			if ('limit' in options) {
+				attrs['db.cf.do.limit'] = options.noCache
+			}
 		}
-		Object.assign(attrs, argArray[0])
 		return attrs
 	},
 	put(argArray) {
 		const attrs = {
 			'db.cf.do.key': argArray[0],
+		} as Record<string, string | number | boolean | undefined>
+
+		if (typeof argArray[0] === 'string') {
+			attrs['db.cf.do.key'] = argArray[0]
+		} else {
+			const keys = Object.keys(argArray[0])
+			attrs['db.cf.do.key'] = keys[0]
+			attrs['db.cf.do.number_of_keys'] = keys.length
 		}
 
-		if (argArray.length > 2) {
-			Object.assign(attrs, argArray[2])
+		const optionsIndex = typeof argArray[1] === 'object' ? 1 : 2
```

Comment:
This doesn't look quite right - `argArray[1]` could be a value with any type supported by the structured clone algorithm. We should instead figure out the index based on if `argArray[0]` is a string or record.
From workers types:
```typescript
put<T>(
  key: string,
  value: T,
  options?: DurableObjectPutOptions
): Promise<void>;
put<T>(
  entries: Record<string, T>,
  options?: DurableObjectPutOptions
): Promise<void>;
```